coderd/database/dbcrypt/dbcrypt.go

```diff
@@ -0,0 +1,196 @@
+	err := db.encryptFields(&params.OAuthAccessToken, &params.OAuthRefreshToken)
+	if err != nil {
+		return database.UserLink{}, err
+	}
+	return db.Store.UpdateUserLink(ctx, params)
+}
+
+func (db *dbCrypt) InsertGitAuthLink(ctx context.Context, params database.InsertGitAuthLinkParams) (database.GitAuthLink, error) {
+	err := db.encryptFields(&params.OAuthAccessToken, &params.OAuthRefreshToken)
+	if err != nil {
+		return database.GitAuthLink{}, err
+	}
+	return db.Store.InsertGitAuthLink(ctx, params)
+}
+
+func (db *dbCrypt) GetGitAuthLink(ctx context.Context, params database.GetGitAuthLinkParams) (database.GitAuthLink, error) {
+	link, err := db.Store.GetGitAuthLink(ctx, params)
+	if err != nil {
+		return database.GitAuthLink{}, err
+	}
+	return link, db.decryptFields(func() error {
+		return db.Store.DeleteGitAuthLink(ctx, database.DeleteGitAuthLinkParams{
+			ProviderID: params.ProviderID,
+			UserID:     params.UserID,
+		})
+	}, &link.OAuthAccessToken, &link.OAuthRefreshToken)
+}
+
+func (db *dbCrypt) UpdateGitAuthLink(ctx context.Context, params database.UpdateGitAuthLinkParams) (database.GitAuthLink, error) {
+	err := db.encryptFields(&params.OAuthAccessToken, &params.OAuthRefreshToken)
+	if err != nil {
+		return database.GitAuthLink{}, err
+	}
+	return db.Store.UpdateGitAuthLink(ctx, params)
+}
+
+func (db *dbCrypt) encryptFields(fields ...*string) error {
+	cipherPtr := db.ExternalTokenCipher.Load()
+	// If no cipher is loaded, then we don't need to encrypt or decrypt anything!
+	if cipherPtr == nil {
+		return nil
+	}
+	cipher := *cipherPtr
+	for _, field := range fields {
+		if field == nil {
+			continue
+		}
+
+		encrypted, err := cipher.Encrypt([]byte(*field))
+		if err != nil {
+			return err
+		}
+		// Base64 is used to support UTF-8 encoding in PostgreSQL.
+		*field = MagicPrefix + base64.StdEncoding.EncodeToString(encrypted)
+	}
+	return nil
+}
+
+// decryptFields decrypts the given fields in place.
+// If the value fails to decrypt, sql.ErrNoRows will be returned.
+func (db *dbCrypt) decryptFields(deleteFn func() error, fields ...*string) error {
+	delete := func(reason string) error {
+		err := deleteFn()
+		if err != nil {
+			return xerrors.Errorf("delete encrypted row: %w", err)
+		}
+		pc, _, _, ok := runtime.Caller(2)
+		details := runtime.FuncForPC(pc)
+		if ok && details != nil {
+			db.Logger.Debug(context.Background(), "deleted row", slog.F("reason", reason), slog.F("caller", details.Name()))
+		}
+		return sql.ErrNoRows
+	}
+
+	cipherPtr := db.ExternalTokenCipher.Load()
+	// If no cipher is loaded, then we don't need to encrypt or decrypt anything!
+	if cipherPtr == nil {
+		for _, field := range fields {
+			if field == nil {
+				continue
+			}
+			if strings.HasPrefix(*field, MagicPrefix) {
+				// If we have a magic prefix but encryption is disabled,
+				// we should delete the row.
+				return delete("encryption disabled")
```

Comment:
What happens when the key is accidentally not specified during coder startup? All rows will start being deleted when they're queried? This seems like a really bad idea

Reply:
All data encrypted at rest is intentionally temporary. We will not do this for any data that must persist.

Reply:
This should be written in a very prominent comment at the top of the file or something then, because I can definitely see this being used for e.g. gitauth by another engineer who doesn't know about this limitation

Reply:
Agreed will add

Reply:
> All data encrypted at rest is intentionally temporary. We will not do this for any data that must persist.

What about ssh keys that should use encryption at rest ?

Reply:
@Adphi that isn't in this PR yet, but I'm not sure what we should do in that case. Maybe stop the server from starting with a warning that is dismissable?

Reply:
If I may, perhaps when the encryption key is wrong, the program should crash loudly to preserve the integrity of the secrets.
And maybe add a command line flag or environment variable to allow all encrypted data to be erased, recovery style.

Reply:
Agreed!